scene_graph/vrd_tutorial.py

```diff
@@ -0,0 +1,335 @@
+
+# %%
+RIDE = 0
+CARRY = 1
+OTHER = 2
+ABSTAIN = -1
+
+
+from snorkel.labeling.lf import labeling_function
+# Category-based LFs
+@labeling_function()
+def LF_ride_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bike", "snowboard", "motorcycle", "horse"]:
+            return RIDE
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_ride_rare_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bus", "truck", "elephant"]:
+            return RIDE
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_carry_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bag", "surfboard", "skis"]:
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_carry_subject(x):
+    if x.object_category == "person":
+        if x.subject_category in ["chair", "bike", "snowboard", "motorcycle", "horse"]:
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_person(x):
+    if x.subject_category != "person":
+        return OTHER
+    return ABSTAIN
+
+
+# Distance-based LFs
+@labeling_function()
+def LF_ydist(x):
+    if x.subject_bbox[3] < x.object_bbox[3]:
+        return OTHER
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_dist(x):
+    if np.linalg.norm(np.array(x.subject_bbox) - np.array(x.object_bbox)) <= 1000:
+        return OTHER
+    return ABSTAIN
+
+
+# Size-based LF
+@labeling_function()
+def LF_area(x):
+    subject_area = (x.subject_bbox[1] - x.subject_bbox[0]) * (
+        x.subject_bbox[3] - x.subject_bbox[2]
+    )
+    object_area = (x.object_bbox[1] - x.object_bbox[0]) * (
+        x.object_bbox[3] - x.object_bbox[2]
+    )
+
+    if subject_area / object_area <= 0.5:
+        return OTHER
+    return ABSTAIN
+
+
+# %% [markdown]
+# Note that the labeling functions have varying empirical accuracies and coverages. Because of the imbalance in how the classes are defined, labeling functions that label the `OTHER` class have higher coverage than labeling functions for `RIDE` or `CARRY`. This reflects the distribution of classes in the dataset as well.
+
+# %%
+from snorkel.labeling.apply import PandasLFApplier
+lfs = [
+    LF_ride_object,
+    LF_ride_rare_object,
+    LF_carry_object,
+    LF_carry_subject,
+    LF_person,
+    LF_ydist,
+    LF_dist,
+    LF_area,
+]
+
+applier = PandasLFApplier(lfs)
+L_train = applier.apply(train_df)
+L_valid = applier.apply(valid_df)
+
+# %%
+from snorkel.labeling.analysis import LFAnalysis
+
+Y_valid = valid_df.label.values
+lf_names = [lf.name for lf in lfs]
+
+lf_analysis = LFAnalysis(L_valid)
+lf_analysis.lf_summary(Y_valid, lf_names=lf_names)
```

Comment:
nit: is it cleaner to do `LFAnalysis(L_valid).lf_summary(Y_valid, lf_names=lf_names)` ?

Reply:
probably but in case users want to run and add their own fields they want to calculate, that usage isn't recommended

Reply:
you're right, sounds god